implementations/rust/ockam/ockam_transport_tcp/src/portal/portal_receiver.rs

```diff
@@ -29,33 +36,38 @@ impl Processor for TcpPortalRecvProcessor {
     type Context = Context;
 
     async fn process(&mut self, ctx: &mut Context) -> Result<bool> {
-        let mut buf = [0u8; BUFFER_SIZE];
-        let len = match self.rx.read(&mut buf).await {
+        self.buf.clear();
+
+        let _len = match self.rx.read_buf(&mut self.buf).await {
             Ok(len) => len,
             Err(err) => {
                 error!("Tcp Portal connection read failed with error: {}", err);
                 return Ok(false);
             }
         };
 
-        if len != 0 {
-            let mut vec = vec![0u8; len];
-            vec.copy_from_slice(&buf[..len]);
-            let msg = PortalInternalMessage::Payload(vec);
-
-            // Let Sender forward payload to the other side
-            ctx.send(route![self.sender_address.clone()], msg).await?;
-
-            Ok(true)
-        } else {
+        if self.buf.is_empty() {
             // Notify Sender that connection was closed
             ctx.send(
                 route![self.sender_address.clone()],
                 PortalInternalMessage::Disconnect,
             )
             .await?;
 
-            Ok(false)
+            return Ok(false);
         }
+
+        // Loop just in case buf was extended (should not happen though)
+        while !self.buf.is_empty() {
+            let at = min(MAX_PAYLOAD_SIZE, self.buf.len());
+            let chunk = self.buf.split_to(at);
```

Comment:
Another fix would to avoid using `split_to` here, since the buffer would be contiguous in that case.